agent/taskresource/cgroup.go

```diff
@@ -0,0 +1,225 @@
+func (c *CgroupResource) GetKnownStatus() CgroupStatus {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	return c.knownStatusUnsafe
+}
+
+// SetCreatedAt sets the timestamp for resource's creation time
+func (c *CgroupResource) SetCreatedAt(createdAt time.Time) {
+	if createdAt.IsZero() {
+		return
+	}
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.createdAt = createdAt
+}
+
+// GetCreatedAt sets the timestamp for resource's creation time
+func (c *CgroupResource) GetCreatedAt() time.Time {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	return c.createdAt
+}
+
+func (c *CgroupResource) Create(task *api.Task) error {
+	err := c.cgroupInit()
+	if err != nil {
+		seelog.Criticalf("unable to setup '/ecs' cgroup: %v", err)
+		return err
+	}
+	err = c.setupTaskCgroup(task)
+	if err != nil {
+		seelog.Criticalf("unable to setup platform resources: %v", err)
+		return err
+	}
+	return nil
+}
+
+func (c *CgroupResource) cgroupInit() error {
+	if c.control.Exists(config.DefaultTaskCgroupPrefix) {
+		seelog.Debugf("Cgroup at %s already exists, skipping creation", config.DefaultTaskCgroupPrefix)
+		return nil
+	}
+	return c.control.Init()
+}
+
+func (c *CgroupResource) setupTaskCgroup(task *api.Task) error {
+	cgroupRoot := c.CgroupRoot
+	seelog.Debugf("Setting up cgroup at: %s for task: %s", cgroupRoot, task.Arn)
+
+	if c.control.Exists(cgroupRoot) {
+		seelog.Debugf("Cgroup at %s already exists, skipping creation", cgroupRoot)
+		return nil
+	}
+
+	linuxResourceSpec, err := task.BuildLinuxResourceSpec()
+	if err != nil {
+		return errors.Wrapf(err, "resource: setup cgroup: unable to build resource spec for task: %s", task.Arn)
+	}
+
+	cgroupSpec := cgroup.Spec{
+		Root:  cgroupRoot,
+		Specs: &linuxResourceSpec,
+	}
+
+	_, err = c.control.Create(&cgroupSpec)
+	if err != nil {
+		return errors.Wrapf(err, "resource: setup cgroup: unable to create cgroup at %s for task: %s", cgroupRoot, task.Arn)
+	}
+
+	// echo 1 > memory.use_hierarchy
+	memoryHierarchyPath := filepath.Join(c.CgroupMountPath, memorySubsystem, cgroupRoot, memoryUseHierarchy)
+	err = c.ioutil.WriteFile(memoryHierarchyPath, enableMemoryHierarchy, rootReadOnlyPermissions)
+	if err != nil {
+		return errors.Wrap(err, "resource: setup cgroup: unable to set use hierarchy flag")
+	}
+
+	return nil
+}
+
+func (c *CgroupResource) Cleanup() error {
+	err := c.control.Remove(c.CgroupRoot)
+	// Explicitly handle cgroup deleted error
+	if err != nil {
+		if err == cgroups.ErrCgroupDeleted {
+			seelog.Warnf("Cgroup at %s has already been removed: %v", c.CgroupRoot, err)
+			return nil
+		}
+		return errors.Wrapf(err, "resource: cleanup cgroup: unable to remove cgroup at %s", c.CgroupRoot)
+	}
+	return nil
+}
+
+// cgroupResourceJSON duplicates CgroupResource fields, only for marshalling and unmarshalling purposes
+type cgroupResourceJSON struct {
+	CgroupRoot      string `json:"CgroupRoot"`
+	CgroupMountPath string `json:"CgroupMountPath"`
+	CreatedAt       time.Time
```

Comment:
i think this needs to be a pointer with omitempty tag. Then, you can set it only if it's been created. 

Also, why do we need this timestamp at all?

Reply:
This could be used for debugging purposes. Also `TaskResource` interface has setters and getters for this field, hence `CgroupResource` has these as well.